File: src/main/java/frc/robot/commands/Autos.java
package frc.robot.commands;

import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.WaitCommand;
import frc.robot.subsystems.superstructure.Superstructure;
import frc.robot.subsystems.swerve.Drivebase;
import java.util.List;
import java.util.function.Supplier;
import lib.autos.AutoRoutine;
import lib.autos.AutoRoutine.AutoAction;
import lib.math.geometry.FieldConstants;
import org.littletonrobotics.junction.networktables.LoggedDashboardChooser;

/** Collection of autonomous routine factories. */
public final class Autos {
  private final Drivebase drivebase;
  private final Superstructure superstructure;

  private final LoggedDashboardChooser<Supplier<Command>> chooser =
      new LoggedDashboardChooser<>("auto");

  private final AutoRoutine ijRoutine;
  private final AutoRoutine i4Routine;
  private final AutoRoutine abcRoutine;
  private final AutoRoutine feRoutine;
  private final AutoRoutine f4Routine;
  private final AutoRoutine idleRoutine;

  public Autos(Drivebase drivebase, Superstructure superstructure) {
    this.drivebase = drivebase;
    this.superstructure = superstructure;

    this.abcRoutine =
        new AutoRoutine(
            List.of(
                new AutoAction(FieldConstants.Reef.Branch.A, FieldConstants.Reef.Level.L4, true),
                new AutoAction(FieldConstants.Reef.Branch.B, FieldConstants.Reef.Level.L2, true),
                new AutoAction(FieldConstants.Reef.Branch.C, FieldConstants.Reef.Level.L3, false),
                new AutoAction(FieldConstants.Reef.Branch.A, FieldConstants.Reef.Level.L3, false)),
            drivebase,
            superstructure);

    this.ijRoutine =
        new AutoRoutine(
            List.of(
                new AutoAction(FieldConstants.Reef.Branch.I, FieldConstants.Reef.Level.L4, true),
                new AutoAction(FieldConstants.Reef.Branch.J, FieldConstants.Reef.Level.L4, true)),
            drivebase,
            superstructure);

    this.i4Routine =
        new AutoRoutine(
            List.of(
                new AutoAction(FieldConstants.Reef.Branch.I, FieldConstants.Reef.Level.L4, true)),
            drivebase,
            superstructure);

    this.feRoutine =
        new AutoRoutine(
            List.of(
                new AutoAction(FieldConstants.Reef.Branch.F, FieldConstants.Reef.Level.L4, true),
                new AutoAction(FieldConstants.Reef.Branch.E, FieldConstants.Reef.Level.L4, true)),
            drivebase,
            superstructure);

    this.f4Routine =
        new AutoRoutine(
            List.of(
                new AutoAction(FieldConstants.Reef.Branch.F, FieldConstants.Reef.Level.L4, true)),
            drivebase,
            superstructure);

    this.idleRoutine = new AutoRoutine(List.of(), drivebase, superstructure);

    chooser.addOption("IJ", ijRoutine::build);
    chooser.addOption("I4", i4Routine::build);
    chooser.addDefaultOption("DRIVE_FORWARDS", this::driveForwards);
  }

  public Command driveForwards() {
    return Commands.sequence(
        Commands.runOnce(() -> drivebase.applyChassisSpeeds(new ChassisSpeeds(-3.0, 0.0, 0.0)), drivebase),
        new WaitCommand(5.0),
        drivebase.getStopCmd());
  }

  public Command getSelectedRoutine() {
    return chooser.get().get();
  }
}

File: src/main/java/frc/robot/commands/Autos.kt
package frc.robot.commands

import com.pathplanner.lib.auto.AutoBuilder
import com.pathplanner.lib.commands.PathPlannerAuto
import com.pathplanner.lib.path.PathPlannerPath
import edu.wpi.first.math.MathUtil
import edu.wpi.first.math.geometry.Pose2d
import edu.wpi.first.math.kinematics.ChassisSpeeds
import edu.wpi.first.math.util.Units
import edu.wpi.first.wpilibj2.command.Command
import edu.wpi.first.wpilibj2.command.Commands
import edu.wpi.first.wpilibj2.command.PrintCommand
import edu.wpi.first.wpilibj2.command.WaitCommand
import frc.robot.commands.swerve.AlignmentCommand
import frc.robot.subsystems.superstructure.Superstructure
import frc.robot.subsystems.superstructure.SuperstructureGoals
import frc.robot.subsystems.swerve.Drivebase
import lib.autos.AutoRoutine
import lib.math.geometry.FieldConstants
import org.littletonrobotics.junction.Logger
import org.littletonrobotics.junction.networktables.LoggedDashboardChooser
import java.util.function.Supplier
import kotlin.jvm.optionals.getOrDefault

class Autos(
    val drivebase: Drivebase,
    val superstructure: Superstructure,
) {

    val IJ_Routine_PP = Supplier {
        val pp_topstart_i = PathPlannerPath.fromPathFile("pp_topstart_i")

        Commands.sequence(
            AutoBuilder.resetOdom(pp_topstart_i.startingHolonomicPose.getOrDefault(Pose2d())),
            AutoBuilder.followPath(pp_topstart_i),
            drivebase.getStopCmd(),
            Commands.waitSeconds(1.0),
            AutoBuilder.followPath(PathPlannerPath.fromPathFile("i_topsource")),
            drivebase.getStopCmd(),
            Commands.waitSeconds(1.0),
            AutoBuilder.followPath(PathPlannerPath.fromPathFile("topsource_j")),
            drivebase.getStopCmd(),
        )
    }

    private val chooser = LoggedDashboardChooser<Supplier<Command>>("auto").apply {
        // addDefaultOption("IDLE", { Commands.sequence(superstructure.getSendToStateCommand { SuperstructureGoals.STOW }, Commands.idle() )})
        addOption("IJ", { IJ_Routine.build() })
        addOption("I4", { I4_Routine.build() })
        addDefaultOption("DRIVE_FORWARDS", {driveForwards()})
//        addOption("FE", { FE_Routine.build() })
//        addOption("F4", { F4_Routine.build() })
//        addOption("B - L2, L3", { bL2_L3() })
//        addOption("idle", { idle_Routine.build()})
    }

    fun driveForwards(): Command {
        return Commands.sequence(
            Commands.runOnce({
                drivebase.applyChassisSpeeds(ChassisSpeeds(-3.0, 0.0, 0.0))
            }, drivebase),
            Commands.waitSeconds(5.0),
            drivebase.getStopCmd()
        )
    }

    val ABC_Routine: AutoRoutine = AutoRoutine(
        listOf(
            Triple(FieldConstants.Reef.Branch.A, FieldConstants.Reef.Level.L4, true),
            Triple(FieldConstants.Reef.Branch.B, FieldConstants.Reef.Level.L2, true),
            Triple(FieldConstants.Reef.Branch.C, FieldConstants.Reef.Level.L3, false),
            Triple(FieldConstants.Reef.Branch.A, FieldConstants.Reef.Level.L3, false),
        ),
        drivebase,
        superstructure,
    )

    val IJ_Routine: AutoRoutine = AutoRoutine(
        listOf(
            Triple(FieldConstants.Reef.Branch.I, FieldConstants.Reef.Level.L4, true),
            Triple(FieldConstants.Reef.Branch.J, FieldConstants.Reef.Level.L4, true),
        ),
        drivebase,
        superstructure,

        )

    val I4_Routine: AutoRoutine = AutoRoutine(
        listOf(
            Triple(FieldConstants.Reef.Branch.I, FieldConstants.Reef.Level.L4, true),
        ),
        drivebase,
        superstructure,

        )

    val FE_Routine: AutoRoutine = AutoRoutine(
        listOf(
            Triple(FieldConstants.Reef.Branch.F, FieldConstants.Reef.Level.L4, true),
            Triple(FieldConstants.Reef.Branch.E, FieldConstants.Reef.Level.L4, true),
        ),
        drivebase,
        superstructure,

        )

    val F4_Routine: AutoRoutine = AutoRoutine(
        listOf(
            Triple(FieldConstants.Reef.Branch.F, FieldConstants.Reef.Level.L4, true),
        ),
        drivebase,
        superstructure,

        )

    val idle_Routine: AutoRoutine = AutoRoutine(
        listOf(
//            Triple(FieldConstants.Reef.Branch.I, FieldConstants.Reef.Level.L4, true),
//            Triple(FieldConstants.Reef.Branch.J, FieldConstants.Reef.Level.L4, true),
        ),
        drivebase,
        superstructure,

        )

    val selectedRoutine: Command
        get() = chooser.get().get()

}